src/main/java/com/mraof/minestuck/alchemy/CardCaptchas.java

```diff
@@ -0,0 +1,231 @@
+package com.mraof.minestuck.alchemy;
+
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+import com.mraof.minestuck.Minestuck;
+import com.mraof.minestuck.item.MSItems;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.util.RandomSource;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.level.levelgen.PositionalRandomFactory;
+import net.minecraftforge.event.server.ServerStartedEvent;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+import net.minecraftforge.fml.common.Mod;
+import net.minecraftforge.registries.ForgeRegistries;
+
+import javax.annotation.Nullable;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Handles the creation of, and look up of, captcha codes for each registered item in the given world.
+ * The captcha is arrived at by hashing the registry name string, which gets some additional randomization via world seed.
+ * Two worlds with the same seed will produce the same captcha for a given registered item, expect in some rare instances where there was hash collision.
+ */
+@Mod.EventBusSubscriber(modid = Minestuck.MOD_ID, bus = Mod.EventBusSubscriber.Bus.FORGE)
+public final class CardCaptchas
+{
+	//TODO use of the character "#" not canonically accurate and suggests there are more punch holes in a captchalogue card then there actually are
+	
+	public static final String AVAILABLE_CHARACTERS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789!?";
+	public static final String EMPTY_CARD_CAPTCHA = "00000000";
+	
+	private static final BiMap<Item, String> REGISTRY_MAP = HashBiMap.create();
+	
+	/**
+	 * Gets the registry name of the item and then returns its captcha or else returns null
+	 */
+	public static String getCaptcha(Item item)
+	{
+		return REGISTRY_MAP.get(item);
+	}
+	
+	@Nullable
+	public static Item getItemFromCaptcha(String captcha)
+	{
+		return REGISTRY_MAP.inverse().get(captcha);
+	}
+	
+	@SubscribeEvent
+	public static void serverStarted(ServerStartedEvent event)
+	{
+		//TODO the second time a world loads, and every subsequent time after that, the captchas are different from the first time the world loads. Both before and after the change, server and client side are synced correctly
+		// Is this still the case, or has a recent change fixed it?
```

Comment:
This should be checked up on.

Reply:
It no longer appears to be the case, will remove!